src/hashtable.c

```diff
@@ -0,0 +1,2133 @@
+            if (compareKeys(ht, data->key, elem_key) == 0) {
+                /* It's a match. */
+                data->state = HASHTABLE_FOUND;
+                return 0;
+            }
+            /* No match. Look for next candidate entry in the bucket. */
+            data->pos++;
+        }
+        /* fall through */
+    case HASHTABLE_NEXT_ENTRY:
+        /* Current bucket is prefetched. Prefetch next potential
+         * matching entry in the current bucket. */
+        if (data->bucket->presence != 0 && data->pos < numBucketPoitions(data->bucket)) {
+            bucket *b = data->bucket;
+            uint8_t h2 = highBits(data->hash);
+            for (int pos = data->pos; pos < numBucketPoitions(b); pos++) {
+                if (isPositionFilled(b, pos) && b->hashes[pos] == h2) {
+                    /* It's a candidate. */
+                    prefetch(b->entries[pos]);
+                    data->pos = pos;
+                    data->state = HASHTABLE_CHECK_ENTRY;
+                    return 1;
+                }
+            }
+        }
+        /* fall through */
+    case HASHTABLE_NEXT_BUCKET:
+        /* Current bucket is prefetched, if any. Find the next bucket in the
+         * chain, or in next table, and prefetch it. */
+        {
+            hashtable *ht = data->hashtable;
+            if (data->bucket == NULL) {
+                data->table = 0;
+                size_t mask = expToMask(ht->bucket_exp[0]);
+                size_t bucket_idx = data->hash & mask;
+                if (ht->rehash_idx >= 0 && bucket_idx < (size_t)ht->rehash_idx) {
+                    /* Skip already rehashed bucket in table 0. */
+                    data->table = 1;
+                    mask = expToMask(ht->bucket_exp[1]);
+                    bucket_idx = data->hash & mask;
+                }
+                data->bucket = &ht->tables[data->table][bucket_idx];
+            } else if (bucketNext(data->bucket) != NULL) {
+                data->bucket = bucketNext(data->bucket);
+            } else if (data->table == 0 && ht->rehash_idx >= 0) {
+                data->table = 1;
+                size_t mask = expToMask(ht->bucket_exp[1]);
+                size_t bucket_idx = data->hash & mask;
+                data->bucket = &ht->tables[data->table][bucket_idx];
+            } else {
+                /* No more tables. */
+                data->state = HASHTABLE_NOT_FOUND;
+                return 0;
+            }
+            prefetch(data->bucket);
+            data->state = HASHTABLE_NEXT_ENTRY;
+            data->pos = 0;
+        }
+        return 1;
+    case HASHTABLE_FOUND:
+        return 0;
+    case HASHTABLE_NOT_FOUND:
+        return 0;
+    }
+    assert(0);
+}
+
+/* Call only when hashtableIncrementalFindStep has returned 0.
+ *
+ * Returns 1 and points 'found' to the entry if an entry was found, 0 if it
+ * was not found. */
+int hashtableIncrementalFindGetResult(hashtableIncrementalFindState *state, void **found) {
```

Comment:
Why not simply return the pointer, with NULL indicating when nothing is found.

Reply:
You can store NULL in the table. If you return NULL, then how do you know if that's the value or if it was not found?

If we allow strong only non-NULL pointers, then we can simplify many of the function prototypes I think. But I wanted to make it generic and not restrict what you can store, as long as it's pointer-sized. You can store longs for example. Maybe it was not worth it though.